// @vanilla-primitives/core/dist/esm/index.js
import Calendar from './calendar.js';
import Dialog from './dialog.js';
import Popover from './popover.js';
import Tabs from './tabs.js';

// Component metadata with correct selectors
const componentMeta = {
  AlertDialog: { selector: '[data-alert-dialog]', Component: Dialog,  options: { isAlert: true } },
  Calendar: { selector: '[data-calendar]', Component: Calendar },
  Dialog: { selector: '[data-dialog]', Component: Dialog },
  Tabs: { selector: '[data-tabs]', Component: Tabs },
  Popover: { selector: '[data-popover]', Component: Popover },
};

// Function to detect and initialize components on the page
function initialize() {
  for (const [name, { selector, Component, options = {} }] of Object.entries(componentMeta)) {
      const elements = document.querySelectorAll(selector);
      if (elements.length > 0) {
          try {
              elements.forEach(el => new Component(el, options));
              console.log(`Initialized ${elements.length} ${name} component(s)`);
          } catch (error) {
              console.error(`Failed to initialize ${name} component:`, error);
          }
      }
  }
}

export { initialize };
