// @vanilla-primitives/core/dist/cjs/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var calendar = require('./calendar.js');
var dialog = require('./dialog.js');
var popover = require('./popover.js');
var tabs = require('./tabs.js');

// Component metadata with correct selectors
const componentMeta = {
  AlertDialog: { selector: '[data-alert-dialog]', Component: dialog,  options: { isAlert: true } },
  Calendar: { selector: '[data-calendar]', Component: calendar },
  Dialog: { selector: '[data-dialog]', Component: dialog },
  Tabs: { selector: '[data-tabs]', Component: tabs },
  Popover: { selector: '[data-popover]', Component: popover },
};

// Function to detect and initialize components on the page
function initialize() {
  for (const [name, { selector, Component, options = {} }] of Object.entries(componentMeta)) {
      const elements = document.querySelectorAll(selector);
      if (elements.length > 0) {
          try {
              elements.forEach(el => new Component(el, options));
              console.log(`Initialized ${elements.length} ${name} component(s)`);
          } catch (error) {
              console.error(`Failed to initialize ${name} component:`, error);
          }
      }
  }
}

exports.initialize = initialize;
