// @vanilla-primitives/core/dist/cjs/tabs.js
'use strict';

class Tabs {
    constructor(element) {
        this.root = element;
        this.tabsList = this.root.querySelector('[data-tabs-list]');
        this.tabsTriggers = Array.from(this.root.querySelectorAll('[data-tabs-trigger]'))
            .filter(trigger => trigger.closest('[data-tabs]') === this.root);
        this.tabsContents = Array.from(this.root.querySelectorAll('[data-tabs-content]'))
            .filter(content => content.closest('[data-tabs]') === this.root);
        this.orientation = this.root.getAttribute('data-tabs-orientation') || 'horizontal';
        this.setupAccessibility();
        this.setupEventListeners();
        this.activateTab(this.tabsTriggers[0], false);
    }

    generateId(prefix) {
        return `${prefix}-${Math.random().toString(36).slice(2, 11)}`;
    }

    setAriaAttributes(element, attrs) {
        Object.entries(attrs).forEach(([key, value]) => element?.setAttribute(key, value));
    }

    setupAccessibility() {
        this.setAriaAttributes(this.tabsList, {
            role: 'tablist',
            'aria-orientation': this.orientation
        });

        this.tabsTriggers.forEach((trigger, index) => {
            const id = this.generateId('tab');
            const panelId = this.generateId('tabpanel');

            this.setAriaAttributes(trigger, {
                id,
                role: 'tab',
                tabindex: index === 0 ? '0' : '-1',
                'aria-selected': index === 0 ? 'true' : 'false',
                'aria-controls': panelId,
                'data-state': index === 0 ? 'active' : ''
            });

            this.setAriaAttributes(this.tabsContents[index], {
                id: panelId,
                role: 'tabpanel',
                tabindex: '0',
                'aria-labelledby': id,
                hidden: index !== 0,
                'data-state': index === 0 ? 'active' : ''
            });
        });
    }

    setupEventListeners() {
        this.tabsTriggers.forEach(trigger => {
            trigger.addEventListener('click', () => this.activateTab(trigger, true));
            trigger.addEventListener('keydown', (e) => this.handleKeyDown(e));
        });
    }

    activateTab(tab, shouldFocus = true) {
        this.tabsTriggers.forEach((t, index) => {
            const isActive = t === tab;
            this.setAriaAttributes(t, {
                'aria-selected': isActive.toString(),
                'tabindex': isActive ? '0' : '-1',
                'data-state': isActive ? 'active' : ''
            });

            const panel = this.tabsContents[index];
            panel.hidden = !isActive;
            panel.setAttribute('data-state', isActive ? 'active' : '');
        });

        if (shouldFocus) {
            tab.focus();
        }
    }

    handleKeyDown(event) {
        const trigger = event.target;
        const triggerList = this.tabsTriggers;
        const index = triggerList.indexOf(trigger);

        const dir = this.orientation === 'vertical' ? {
            ArrowUp: -1,
            ArrowDown: 1
        } : {
            ArrowLeft: -1,
            ArrowRight: 1
        };

        if (dir[event.key] !== undefined) {
            event.preventDefault();
            const newIndex = (index + dir[event.key] + triggerList.length) % triggerList.length;
            this.activateTab(triggerList[newIndex], true);
        } else if (event.key === 'Home') {
            event.preventDefault();
            this.activateTab(triggerList[0], true);
        } else if (event.key === 'End') {
            event.preventDefault();
            this.activateTab(triggerList[triggerList.length - 1], true);
        }
    }
}

module.exports = Tabs;
